extensions/wikia/Recirculation/js/incontent.js

```diff
@@ -0,0 +1,187 @@
+/*global require*/
+require([
+	'jquery',
+	'wikia.window',
+	'wikia.abTest',
+	'wikia.nirvana',
+	'wikia.mustache',
+	'ext.wikia.recirculation.tracker',
+	'ext.wikia.recirculation.utils'
+], function ($, w, abTest, nirvana, Mustache, tracker, utils) {
+	// Currently only showing for English communities
+	if (w.wgContentLanguage !== 'en') { return; }
+
+
+	var $container = $('#mw-content-text'),
+		template;
+
+	function injectInContentWidget($container) {
+		var $links = $container.find('a'),
+			sections = buildSections($container.find('h2')),
+			width,
+			firstSuitableSection,
+			topTitles;
+
+		// If this page doesn't have enough content (either links or sections) we
+		// don't want to show this widget
+		if ($links.length < 8 || sections.length < 3) {
+			return;
+		}
+
+		// The idea is to show links above the first section under an infobox
+		width = $container.outerWidth();
+		firstSuitableSection = sections.find(function(item, index) {
+			return item.width === width;
+		});
+
+		if (!firstSuitableSection) {
+			return;
+		}
+
+		topTitles = findTopTitles($links);
+
+		if (topTitles.length < 3) {
+			return;
+		}
+
+		nirvana.sendRequest({
+			controller: 'ArticlesApi',
+			method: 'getDetails',
+			format: 'json',
+			type: 'get',
+			data: {
+				titles: topTitles.join(','),
+				abstract: 0,
+				width: 50,
+				height: 50
+			},
+			callback: renderArticles(firstSuitableSection)
+		});
+	}
+
+	function renderArticles(section) {
+		var placeholderImage = w.wgExtensionsPath + '/wikia/Recirculation/images/placeholder.png';
+		tracker.trackImpression('in-content');
+
+		return function (response) {
+			var items = [],
+				html;
+
+			$.each(response.items, function(index, item) {
+				item.thumbnail = item.thumbnail || placeholderImage;
+				items.push(item);
+			});
+
+			$html = $(Mustache.render(template, {
+				title: $.msg('recirculation-incontent-title'),
+				items: items
+			}));
+
+			section.$start.before($html);
+
+			$html.on('mousedown', 'a', function() {
+				tracker.trackClick('in-content');
+			});
+		}
+	}
+
+	/**
+	 * Build the sections array
+	 *
+	 * DOM/layout querying: OK
+	 * DOM modification:    forbidden
+	 *
+	 * @param {jQuery} $headers headers dividing the article to sections
+	 * @returns {Section[]}
+	 */
+	function buildSections($headers) {
```

Comment:
This method [look familiar](https://github.com/Wikia/app/blob/6f82a77805c4eb66f198e43632ae1e40b81209a3/extensions/wikia/AdEngine/js/slot/inContent.js#L98), does it @gabrys? :japanese_ogre: I'm pinging Rychu since those methods are similar but not exactly the same. So, it's a kind of red flag in my head which I don't know how to remove right now. If the methods were the same I'd suggest creating for example `wikia.sectionsBuilder` module and re-use its method in those two places :stuck_out_tongue_winking_eye: